### ISSUE TITLE
setting activeStandby: false on existing CephFilesystem has no effect

### ISSUE THREAD

**Issue by jhoblitt** [rook/rook#12576]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Changing an existing `CephFilesystem` from `activeStandby: true` to `activeStandby: false` has no effect.  The standby-replay mds have to be manually disabled by running `ceph fs set <fs> allow_standby_replay false` from the toolbox.

**Expected behavior:**

The expectation is that rook would disabled `standby-replay` mode.

**How to reproduce it (minimal and precise):**

Create a cephfs

```yaml
apiVersion: ceph.rook.io/v1
kind: CephFilesystem
metadata:
  name: project
  namespace: rook-ceph
spec:
  metadataPool:
    failureDomain: host
    replicated:
      size: 3
    quotas:
      maxSize: 10Gi
  dataPools:
    - failureDomain: host
      replicated:
        size: 3
      quotas:
        maxSize: 500Gi
  metadataServer:
    activeCount: 3
    activeStandby: true
    resources:
      limits:
        cpu: "4"
        memory: 2Gi
      requests:
        cpu: "4"
        memory: 2Gi
  preserveFilesystemOnDelete: false
```

<allow rook/ceph to settle>

Update the cephfs to have `activeStandby: false`:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephFilesystem
metadata:
  name: project
  namespace: rook-ceph
spec:
  metadataPool:
    failureDomain: host
    replicated:
      size: 3
    quotas:
      maxSize: 10Gi
  dataPools:
    - failureDomain: host
      replicated:
        size: 3
      quotas:
        maxSize: 500Gi
  metadataServer:
    activeCount: 3
    activeStandby: false
    resources:
      limits:
        cpu: "4"
        memory: 2Gi
      requests:
        cpu: "4"
        memory: 2Gi
  preserveFilesystemOnDelete: false
```

<wait for rook to settle>

Check status of said cephfs, which shows mds still in `standby-replay` mode.

```bash
[root@rook-ceph-tools-7cf684f75c-kr2b8 /]# ceph fs status project
project - 1 clients
=======
RANK      STATE          MDS        ACTIVITY     DNS    INOS   DIRS   CAPS  
 0        active      project-a  Reqs:    0 /s  10.9k  4854     32     10   
 1        active      project-b  Reqs:    0 /s  11.1k    28     17      2   
 2        active      project-c  Reqs:    0 /s  70.4k  63.1k   524      0   
0-s   standby-replay  project-f  Evts:    0 /s  10.9k  4853     32      0   
1-s   standby-replay  project-d  Evts:    0 /s  11.1k    28     17      0   
2-s   standby-replay  project-e  Evts:    0 /s  70.4k  63.1k   514      0   
      POOL          TYPE     USED  AVAIL  
project-metadata  metadata  1428M  4861G  
 project-data0      data     583G  4861G  
STANDBY MDS  
  auxtel-e   
  auxtel-b   
  auxtel-f   
MDS version: ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
```

**Environment**:

* Rook version (use `rook version` inside of a Rook Pod):
`1.11.2`

**Comment by sp98** [comment]
Previous issue around this got closed - https://github.com/rook/rook/issues/9857

**Comment by travisn** [comment]
@jhoblitt You only see this issue when editing the filesystem CR, right? When a new filesystem is created with the standby disabled, it should be working. So we need to fix this when editing the CR.

**Comment by jhoblitt** [comment]
@travisn That is a good question... I'm unsure if I ever tested creating a new cephfs with standby disabled.  For context, I was disabling standby-replay on some production systems due to a mds memory leak for standby-replay instances.

**Comment by BlaineEXE** [comment]
I thought when I tested it, it didn't matter whether it was after editing or on a new creation, active standby MDSes were always deployed no matter what.

**Comment by jhoblitt** [comment]
I think this is still a live concern.

**Comment by R-omk** [comment]
still relevant,    
https://github.com/rook/rook/issues/9857#issuecomment-1061845459

parasitic deployments are created even after manual removal

@travisn  please reopen issue

**Comment by travisn** [comment]
@R-omk Do you mean #9857 should be reopened? Please comment on that issue with more details. Not sure what you mean by parasitic deployments.

**Comment by R-omk** [comment]
@travisn ,   I investigated the problem, look at the logs in my screenshot, this is the scenario in which I reduced the number from three to two.

The logs show that the actual decrease occurred after waiting and receiving an error, however, from the waiting code (https://github.com/rook/rook/blob/460fcda544df8a6acf99afb8e052363c3ac86e2b/pkg/daemon/ceph/client/filesystem.go#L250) it is clear that the current value is compared with the actual one, but the current value will always be equal to three, and two is expected.

![image](https://github.com/rook/rook/assets/1633879/bb712dd3-1f3b-4473-b95b-f7111554b473)

**Comment by R-omk** [comment]
I also found out why turning off  `activeStandby: true`  - >   `activeStandby: false` does not reduce the number of deployments.   

.. there's just hardcoded  increasing twice

https://github.com/rook/rook/blob/460fcda544df8a6acf99afb8e052363c3ac86e2b/pkg/operator/ceph/file/mds/mds.go#L128

**Comment by BlaineEXE** [comment]
@R-omk Can you copy the full text of the log instead of just a screenshot. I can't see the base error to debug the issue further. 

It seems that the code is starting to wait for the num active mdses to change, but then exits immediately. Perhaps this is an issue with the context in `PollUntilContextTimeout`. 🤔

**Comment by visdmin** [comment]
Still relevant 👍🏼 

Just wasted multiple hours trying to get rid of the extra mds deployment using rook-ceph-cluster Helm chart values :
```yaml
metadataServer:
  activeCount: 1
  activeStandby: false
``` 

I am new to this project, but as fresh user I can still state that this behaviour should be at least well documented !

**Comment by Termibg22** [comment]
I’m facing the same problem.
When I first set up the filesystem I used:

```
metadataServer:
  activeCount: 1
  activeStandby: true
```

This correctly launched two metadata servers. But after disabling that parameter nothing changes, the extra metadata server remains deployed.

If I deploy a new filesystem from the beginning with:

```
metadataServer:
  activeCount: 1
  activeStandby: false
```

it still deploys two metadata servers.

I believe this issue should be reopened if possible @BlaineEXE .

**Comment by jhoblitt** [comment]
@Termibg22 @visdmin Could you report what rook/ceph version your seeing the regression with?

**Comment by Termibg22** [comment]
> [@Termibg22](https://github.com/Termibg22) [@visdmin](https://github.com/visdmin) Could you report what rook/ceph version your seeing the regression with?

I observed this behavior with Rook v1.8.0 and Ceph v19.2.3. I tried upgrading the Operator to v1.8.1 and then to v1.8.2 but the issue persisted.

These are the logs from Rook v1.8.2 with Ceph v19.2.3 if I deploy the filesystem first with `activeStandby: true` and I later change it to `activeStandby: false`.

````
2025-09-15 19:18:49.974231 I | ceph-file-controller: start running mdses for filesystem "filesystem"
2025-09-15 19:18:50.149201 I | cephclient: getting or creating ceph auth key "mds.filesystem-a"
2025-09-15 19:18:50.344403 I | op-mds: setting mds config flags
2025-09-15 19:18:50.344421 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:18:50.513503 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:18:50.521256 I | cephclient: getting or creating ceph auth key "mds.filesystem-b"
2025-09-15 19:18:51.520922 I | op-mds: setting mds config flags
2025-09-15 19:18:51.520948 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:18:51.690240 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:18:55.913781 I | cephclient: deleting ceph auth "client.bootstrap-mds"
2025-09-15 19:18:56.081231 I | ceph-cluster-controller: successfully deleted bootstrap key "client.bootstrap-mds" in the namespace "rook-ceph"
2025-09-15 19:19:18.273757 I | cephclient: deleting ceph auth "client.bootstrap-mds"
2025-09-15 19:19:18.451910 I | ceph-cluster-controller: successfully deleted bootstrap key "client.bootstrap-mds" in the namespace "rook-ceph"
2025-09-15 19:21:34.567848 I | op-config: deleting "mds.filesystem-a" "bdev_flock_retry" option from the mon configuration database
2025-09-15 19:21:34.740368 I | op-mds: successfully deleted mds config for "mds.filesystem-a" in mon configuration database
2025-09-15 19:21:34.740373 I | cephclient: deleting ceph auth "mds.filesystem-a"
2025-09-15 19:21:34.942409 I | op-mds: successfully deleted mds CephX key for "mds.filesystem-a"
2025-09-15 19:21:35.081244 I | op-config: deleting "mds.filesystem-b" "bdev_flock_retry" option from the mon configuration database
2025-09-15 19:21:35.227816 I | op-mds: successfully deleted mds config for "mds.filesystem-b" in mon configuration database
2025-09-15 19:21:35.227824 I | cephclient: deleting ceph auth "mds.filesystem-b"
2025-09-15 19:21:35.405017 I | op-mds: successfully deleted mds CephX key for "mds.filesystem-b"
2025-09-15 19:21:40.716135 I | ceph-spec: object "rook-ceph-mds-filesystem-a-keyring" matched on delete, reconciling
2025-09-15 19:21:40.717203 I | ceph-spec: object "rook-ceph-mds-filesystem-b" matched on delete, reconciling
2025-09-15 19:21:40.717374 I | ceph-spec: object "rook-ceph-mds-filesystem-b-keyring" matched on delete, reconciling
2025-09-15 19:21:40.717544 I | ceph-spec: object "rook-ceph-mds-filesystem-a" matched on delete, reconciling
2025-09-15 19:22:18.945950 I | ceph-file-controller: start running mdses for filesystem "filesystem"
2025-09-15 19:22:19.120408 I | cephclient: getting or creating ceph auth key "mds.filesystem-a"
2025-09-15 19:22:19.322298 I | op-mds: setting mds config flags
2025-09-15 19:22:19.322365 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:22:19.472057 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:22:19.476903 I | cephclient: getting or creating ceph auth key "mds.filesystem-b"
2025-09-15 19:22:19.704786 I | op-mds: setting mds config flags
2025-09-15 19:22:19.704809 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:22:19.876128 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:22:47.663040 I | ceph-file-controller: start running mdses for filesystem "filesystem"
2025-09-15 19:22:47.832055 I | cephclient: getting or creating ceph auth key "mds.filesystem-a"
2025-09-15 19:22:48.029384 I | op-mds: setting mds config flags
2025-09-15 19:22:48.029521 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:22:48.182828 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:22:48.188447 I | op-mds: deployment for mds "rook-ceph-mds-filesystem-a" already exists. updating if needed
2025-09-15 19:22:48.192267 I | op-k8sutil: deployment "rook-ceph-mds-filesystem-a" did not change, nothing to update
2025-09-15 19:22:48.192284 I | cephclient: getting or creating ceph auth key "mds.filesystem-b"
2025-09-15 19:22:48.368067 I | op-mds: setting mds config flags
2025-09-15 19:22:48.368085 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:22:48.514959 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:22:48.521274 I | op-mds: deployment for mds "rook-ceph-mds-filesystem-b" already exists. updating if needed
2025-09-15 19:22:48.524722 I | op-k8sutil: deployment "rook-ceph-mds-filesystem-b" did not change, nothing to update
````

These are the logs from Rook v1.8.2 with Ceph v19.2.3 if I deploy the filesystem first with `activeStandby: false`:

````
2025-09-15 19:18:49.974231 I | ceph-file-controller: start running mdses for filesystem "filesystem"
2025-09-15 19:18:50.149201 I | cephclient: getting or creating ceph auth key "mds.filesystem-a"
2025-09-15 19:18:50.344403 I | op-mds: setting mds config flags
2025-09-15 19:18:50.344421 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:18:50.513503 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-a") to the mon configuration database
2025-09-15 19:18:50.521256 I | cephclient: getting or creating ceph auth key "mds.filesystem-b"
2025-09-15 19:18:51.520922 I | op-mds: setting mds config flags
2025-09-15 19:18:51.520948 I | op-config: setting option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:18:51.690240 I | op-config: successfully set option "mds_join_fs" (user "mds.filesystem-b") to the mon configuration database
2025-09-15 19:18:55.913781 I | cephclient: deleting ceph auth "client.bootstrap-mds"
2025-09-15 19:18:56.081231 I | ceph-cluster-controller: successfully deleted bootstrap key "client.bootstrap-mds" in the namespace "rook-ceph"
2025-09-15 19:19:18.273757 I | cephclient: deleting ceph auth "client.bootstrap-mds"
2025-09-15 19:19:18.451910 I | ceph-cluster-controller: successfully deleted bootstrap key "client.bootstrap-mds" in the namespace "rook-ceph"
````

**Comment by travisn** [comment]
@Termibg22 Did you mean you tested on the latest Rook v1.18.x? 

The behavior is certainly observed in my testing that 2x mds daemons are always created. 

In the comments, the intention for `activeStandby` appeared to be:
```
# If false, standbys will be available, but will not have a warm cache.
```

I remember now this was the intention, but now I'm questioning why you would want a standby mds daemon that does not have an updated cache. Why wouldn't you enable the activeStandy if you want another mds daemon? Requiring the mds standby seems unnecessary.

**Comment by Termibg22** [comment]
> [@Termibg22](https://github.com/Termibg22) Did you mean you tested on the latest Rook v1.18.x?

That's right, I tested it with the latest version that was available at the time, v1.18.2. Sorry, I wrote the version incorrectly.

> 
> The behavior is certainly observed in my testing that 2x mds daemons are always created.
> 
> In the comments, the intention for `activeStandby` appeared to be:
> 
> ```
> # If false, standbys will be available, but will not have a warm cache.
> ```
> 
> I remember now this was the intention, but now I'm questioning why you would want a standby mds daemon that does not have an updated cache. Why wouldn't you enable the activeStandy if you want another mds daemon? Requiring the mds standby seems unnecessary.

I think you're right and after reviewing the PR I believe the proposed behavior is much more consistent with the parameter name and with what users can expect from the functions. It wasn’t very intuitive that deploying a single MDS daemon wasn’t possible.

If I understood correctly the new behavior I would update the parameter definition in the documentation to something like this:

```
activeStandby: Whether each active MDS instance will have a warm cache standby MDS instance. When set to true, twice the number of MDS daemons defined by activeCount will be deployed. When set to false, only the activeCount MDS daemons will be deployed and no warm standby MDS will be created.
```

Thank you for your work
